python/psautohint/otfFont.py

```diff
@@ -456,35 +455,69 @@ def build_hint_order(hints):
     return hints, hint_order
 
 
-RefHintMask = namedtuple("RefHintMask", "pos hbytes")
-
-
-class MMHintInfo:
-    def __init__(self, name=None):
-        self.reset(name)
+def make_abs(hint_pair):
+    bottom_edge, delta = hint_pair
+    new_hint_pair = [bottom_edge, delta]
+    if delta in [-20, -21]:
+        new_hint_pair[0] = bottom_edge + delta
+        new_hint_pair[1] = bottom_edge
+    else:
+        new_hint_pair[1] = bottom_edge + delta
+    return new_hint_pair
+
+
+def check_hint_overlap(hint_list, last_idx, bad_hint_idxs):
+    # return True if there is an overlap.
+    prev = hint_list[0]
+    for i, hint_pair in enumerate(hint_list[1:], 1):
+        if prev[1] >= hint_pair[0]:
+            bad_hint_idxs.add(i + last_idx)
+        prev = hint_pair
+
+
+def check_hint_pairs(hint_pairs, mm_hint_info, last_idx=0):
+    # pairs must be in ascending order by bottom (or left) edge,
+    # and pairs in a hint group must not overlap.
+
+    # check order first
+    hint_list = [make_abs(hint_pair) for hint_pair in hint_pairs]
+    bad_hint_idxs = set()
+    prev = hint_list[0]
+    for i, hint_pair in enumerate(hint_list[1:], 1):
```

Comment:
When a pair of stem hints is not in the ascending order, it may not be so trivial to determine which one should be regarded as bad. The code here appears to mark the second one in a pair as bad, but consider four stem hints: A, B, C, D in the correct order, then given a list of: C, A, B, D. I believe this code will regard A and B as bad, but C as bad would minimize the number of bad hints. The same could be said to overlapping hints.

Reply:
@blueshade7 I see your point. Looking at several cases, I think that always removing the first pair will not cause this problem - do you agree?

Reply:
I cannot be sure without seeing real world examples. Let me assume now it is doing the right thing or at least not a bad thing. Let's revisit when we see a problem later.